### DISCUSSION TITLE
Additional layer between <widgets> and <features> in big projects

### DISCUSSION THREAD

**Discussion by stsiushkevich** [feature-sliced/documentation#621] [Ideas]
I'm working on a big project and from time to tome facing with a requirement of additional level, asa ```microwidgets``` , between ```widgets``` and ```features``` to avoid any logic in ```pages``` level. Simple example: feature ```EventForm``` can be wrapped into popup ```EventEditor``` with some logic, like as confirmation dialog or something else. By FSD ```EventEditor``` should be widget, so that it can be used only on ```pages```, what can cause logic in page. Some intermediate layer can help overcome such problem.

**Comment by illright** [comment]
Why do you want to avoid logic on pages?

**Reply by stsiushkevich** [reply]
![image](https://github.com/feature-sliced/documentation/assets/12151060/779f6158-920a-4e79-ad06-288a0eceaf44)

I agreed with the idea, that page can only be composed from other lower layers, but should avoid any logic, because the logic is low level thing, which should be incapsulated in widgets. If allow logic on page there can be logic from multiple widgets and it cause to pollution of page.

**Comment by illright** [comment]
What layer would you propose? Each of the current layers is characterized not only by its position in the layer stack, but also by its unique semantics:

- shared — non-business code
- entities — business entities
- features — actions on entities for business value
- widgets — composition into meaningful blocks on the screen
- pages — composition into pages
- app — infrastructure code

To introduce a new layer, we first need a name and semantic value.

**Reply by stsiushkevich** [reply]
I have no idea about it, but I discovered a need some intermediate level to incapsulate a part of business logic. It doesn't brake any conception of FSD. Furthermore, can help to minify widgets in big projects.

**Reply by illright** [reply]
In my experience, all instances of _another layer needed_ were resolvable by rethinking the decomposition. I encourage you to give that a try, asking help from the Telegram/Discord community if needed.

Since me and the rest of the core team haven't felt the need to add a new layer, it will only happen if you lead the initiative and create a convincing proposal.
